src/reconcile/rec_col.c

```diff
@@ -310,44 +318,519 @@ __wt_rec_col_int(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_REF *pageref)
+            /* Update for skipped item. Shouldn't happen, but just in case it does, skip it. */
+            continue;
+
+        /* Copy in all the preexisting time windows for keys before this one. */
+        while (tw < numtws && origstartrecno + page->pg_fix_tws[tw].recno_offset < recno) {
+            /* Get the previous time window so as to copy it. */
+            cell = WT_COL_FIX_TW_CELL(page, &page->pg_fix_tws[tw]);
+            __wt_cell_unpack_kv(session, page->dsk, cell, &unpack);
+
+            /* If it's from a previous run, it might become empty; if so, skip it. */
+            if (!WT_TIME_WINDOW_IS_EMPTY(&unpack.tw))
+                WT_ERR(__wt_rec_col_fix_addtw(session, r,
+                  (uint32_t)(origstartrecno + page->pg_fix_tws[tw].recno_offset - curstartrecno),
+                  &unpack.tw));
+            tw++;
+        }
+
+        /*
+         * Fake up an unpack record to pass to update selection; it needs to have the current
+         * on-disk value and its timestamp, if any, and it also needs to be tagged as a value cell.
+         * This is how that value gets into the history store if that's needed.
+         */
+        if (tw < numtws && origstartrecno + page->pg_fix_tws[tw].recno_offset == recno) {
+            /* Get the on-disk time window by unpacking the value cell. */
+            cell = WT_COL_FIX_TW_CELL(page, &page->pg_fix_tws[tw]);
+            __wt_cell_unpack_kv(session, page->dsk, cell, &unpack);
+        } else {
+            /* Fake up a value cell with a default time window. */
+            unpack.type = WT_CELL_VALUE;
+            WT_TIME_WINDOW_INIT(&unpack.tw);
+        }
+
+        /*
+         * Stick in the current on-disk value. We can't use __bit_getv_recno here because it
+         * implicitly uses pageref->ref_recno to figure the offset; that's wrong if salvage has
+         * changed the page origin.
+         */
+        WT_ASSERT(session, page->dsk != NULL && origstartrecno != WT_RECNO_OOB);
+        val = __bit_getv(page->pg_fix_bitf, recno - origstartrecno, btree->bitcnt);
+        unpack.data = &val;
+        unpack.size = 1;
+
+        WT_ERR(__wt_rec_upd_select(session, r, ins, NULL, &unpack, &upd_select));
         upd = upd_select.upd;
-        if (upd != NULL)
-            __bit_setv(
-              r->first_free, WT_INSERT_RECNO(ins) - pageref->ref_recno, btree->bitcnt, *upd->data);
+        if (upd == NULL) {
+            /*
+             * It apparently used to be possible to get back no update but a nonempty time window to
+             * apply to the current on-disk value. As of Oct. 2021 this is no longer the case;
+             * instead we get back an update with a copy of the current on-disk value. In case of
+             * future changes, assert that there's nothing to do.
+             */
+            WT_ASSERT(session, WT_TIME_WINDOW_IS_EMPTY(&upd_select.tw));
+            continue;
```

Comment:
I put that assert in because not having it led me on a lengthy wild-goose chase. Admittedly, that was based partly on an obsolete or maybe just misleading comment in rec_visibility.c that I've since updated, but I'd still kind of rather keep the assertion. If you really want it gone I can get rid of it though...

Reply:
I can live with it. :)